pkg/openstack/keystone.go

```diff
@@ -17,17 +17,19 @@ import (
 
 // ReconcileKeystoneAPI -
 func ReconcileKeystoneAPI(ctx context.Context, instance *corev1beta1.OpenStackControlPlane, helper *helper.Helper) (ctrl.Result, error) {
-	if !instance.Spec.Keystone.Enabled {
-		return ctrl.Result{}, nil
-	}
-
 	keystoneAPI := &keystonev1.KeystoneAPI{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      "keystone", //FIXME (keystone doesn't seem to work unless named "keystone")
 			Namespace: instance.Namespace,
 		},
 	}
 
+	if !instance.Spec.Keystone.Enabled {
+		if res, err := CheckDeleteDisabled(ctx, helper, keystoneAPI); err != nil {
```

Comment:
Deleting Keystone while still having other services using it (like cinder) will not work as KeystoneService and KeystoneEndpoint CRs will be blocked by the service finalizers. I think this is why we have the staged delete logic in https://github.com/openstack-k8s-operators/openstack-operator/blob/f2d183dd6292dcca4050e59ca3d7bd6cc5534545/controllers/core/openstackcontrolplane_controller.go#L263-L282 So I think we need some check here too to avoid this situation.

Reply:
IIUC you're talking about service dependencies. I think we should do that in the validation webhook and keep the lifecycle logic simple for services. I guess the commit message talks about it.

Reply:
Yepp it is about dependencies between the services. I.e. keystone cannot be deleted while nova is actively using KeystoneService CRs. 

Hm. So a validation webhook can reject a keystone disable request if nova is enabled. That part is good. And I guess that will be implemented in `ValidateServices` above. 

What if an single update is coming with enabled=False for both keystone and nova (and assuming no other keystone dependencies is enabled)? Will it be rejected with an error than nova needs to be disabled first? Or will it be allowed? 

If the latter then the edge case, where the update request contains enabled=False for all the services is logically equivalent for deleting all sub resources during the OpenStackControl plane delete. So I think the two delete scenario ( 1) disable all services 2) delete OpenStackControlPlane CR) has the same dependency handling needs. So I'm wondering if we could reuse parts of `reconcileDelete` for the disable service code path too.

Hm, even if we do the former then two update requests can come in closely after each other (nova disabled first then keystone) making it possible that nova is not deleted (even if it is already disabled) when keystone is disabled.

Reply:
> What if an single update is coming with enabled=False for both keystone and nova 
That would be allowed by the webhook.

> where the update request contains enabled=False for all the services
That's a simple scenario where we would just allow it in the webhook without checking dependencies. We can make the validations as simple/complicated as we want.

> Hm, even if we do the former then two update requests can come in closely after each other
I assume the requests would be queued one after the other for the OpenStackControlPlane CR and would be handled accordingly and we should not have any issues right?

Reply:
> > where the update request contains enabled=False for all the services
> 
> That's a simple scenario where we would just allow it in the webhook without checking dependencies. We can make the validations as simple/complicated as we want.

OK, but that would mean delete all the subresources. Which is very similar to the case when OpenStackControlPlane is deleted. So either the code in `reconcileDelete` is over-complicated for that scenario or the the currently proposed independent per service delete is too simplistic for it. @abays you authored the controller deletion of subresources in `reconcileDelete` so I guess you have a better view on it.

Reply:
> OK, but that would mean delete all the subresources

I don't think we need to. To me it looks like CheckDeleteSubresources is doing things we don't need to and can be simplified.  reconcileDelete of service instances would called when you make a delete api call and that should be sufficient.

Reply:
Currently, you can delete (or, I suppose, disable) any of these [1] resources in any order you please.  What's important in the teardown logic, however, is that Keystone [2] is then removed only _after_ those resources have been deleted.  And finally, after Keystone and the rest are gone, you are free to remove MariaDB [3].  If you don't at least follow the order of [1] and then [2], the finalizers on the Keystone resources will prevent tear-down from completing.  

Keep in mind that this logic referenced here (and in above comments) is only for the _complete_ teardown of the `OpenStackControlPlane`.  Deleting individual services would be different, but any attempt to disable the Keystone service (which would entail deleting `KeystoneAPI`, `KeystoneService`, `KeystoneEndpoints`, etc, I am assuming) would be blocked by the finalizers on the Keystone service if other OpenStack services exist and have placed finalizers on the Keystone service resources.  And of course disabling and deleting the MariaDB service will torpedo everything.  For whatever reason MariaDB currently does not have the same level of dependency integrity protection via finalizers that Keystone does (it has them on `MariaDBDatabase`, but not on `MariaDB`, which can be deleted and hose everything).

[1] https://github.com/openstack-k8s-operators/openstack-operator/blob/534f09255ae43a7e7c48e0d118d81d37bed50380/controllers/core/openstackcontrolplane_controller.go#L267-L274
[2] https://github.com/openstack-k8s-operators/openstack-operator/blob/534f09255ae43a7e7c48e0d118d81d37bed50380/controllers/core/openstackcontrolplane_controller.go#L278
[3] https://github.com/openstack-k8s-operators/openstack-operator/blob/534f09255ae43a7e7c48e0d118d81d37bed50380/controllers/core/openstackcontrolplane_controller.go#L281

Reply:
> Deleting individual services would be different, but any attempt to disable the Keystone service (which would entail deleting KeystoneAPI, KeystoneService, KeystoneEndpoints, etc, I am assuming) would be blocked by the finalizers on the Keystone service if other OpenStack services exist and have placed finalizers on the Keystone service resources.

Yeah looks like we add a finalizer[1] on the KeystoneAPI for the service instance and that won't let the KeystoneAPI  disable/delete attempt to succeed (would wait) unless the dependent service instances are disabled/deleted.  We probably need similar level of protection with MariaDB/MariaDBDatabase. I'll see if I can add that.

And you're right we need to do them in ordered fashion as you've implemented when tearing down OpenStackControlPlane instance, else api call to delete service instances won't return and the whole teardown would get stuck. 

[1] https://github.com/openstack-k8s-operators/keystone-operator/blob/master/controllers/keystoneservice_controller.go#L181-L190

Reply:
I've proposed a patch go add finalizes[1] on MariaDB for services, so that it can't be deleted before the services. It It It would also require rbac changes in all operators. I've patch for keystone-operator[2] already proposed.

[1] https://github.com/openstack-k8s-operators/lib-common/pull/165
[2] https://github.com/openstack-k8s-operators/keystone-operator/pull/186